Add support for memory mappings with large pages

lexprfuncall opened the pull request:
This change extends the memory managers in Erlang/OTP to
optionally map virtual memory on Linux using large pages.  Large
pages make more efficient use of the TLB by reducing the dynamic
frequency of TLB misses thereby increasing application
performance.  The effect or reducing TLB misses can be dramatic.
On our server workloads we have observed >10% performance
improvements simply from enabling mappings with large pages.

The opportunity for such an improvement comes from the growing
disparity between the size of the TLB and main memory.  The
typical 64-bit x86 has a 64-entry L1 data TLB and a 128-entry L1
instruction TLB, addressing 256KiB and 512KiB of data and text
respectively, at 4KiB per entry.  This is 4-5 orders of magnitude
smaller than the amount of memory installed in the typical server
of today.  When the TLB is missed the processor must perform an
address translate by traversing a multi-level table in main
memory, a costly operation.  This overhead is largely dispensable
by using a larger page size.  For mappings using a 2MiB or 1GiB
page in lieu of a 4KiB page the L1 TLB can address 32MiB or 4GiB,
respectively, a more meaningful fraction of typical working sets.

To take advantage of large pages in Erlang, we identified three
classes of virtual memory allocations that would benefit from
large pages: the .text segment, heap allocations, JIT
allocations.  While there are many strategies for using large
pages on Linux, we chose to use Transparent Huge Pages which are
the most flexible.  Briefly put, in order to use THP we arranged
for all of these classes of virtual memory allocations to be
aligned and sized to a multiple of a large page and performed a
system call that advised the kernel to use large pages for the
mapping.

While this change is Linux specific, for the most part, other
operating systems have similar, and sometimes better, mechanisms
for achieving the same effect.  We believe it is be possible to
use large pages for heap and JIT allocations on FreeBSD, Linux,
Solaris, and Windows.  The changes to AsmJit in particular
already contain work to this effect.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/erlang/otp?pullRequest=7977) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/erlang/otp?pullRequest=7977) before we can accept your contribution.<br/><hr/>**lexprfuncall** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/erlang/otp?pullRequest=7977) it.</sub>

jhogberg commented on erts/emulator/beam/erl_init.c:
```diff
@@ -762,6 +806,110 @@ static void ethr_ll_free(void *ptr)
     erts_free(ERTS_ALC_T_ETHR_LL, ptr);
 }
 
+#if defined(__linux__) && defined(__x86_64__) && defined(__LP64__)
+static int
+is_transparent_hugepage_madvise_enabled(void)
```
This feels a bit too platform-specific to have in `erl_init.c`, can you move the meat of it to `erts/emulator/sys/unix/sys_huge_pages.c` or similar?

lexprfuncall replied:
Sure, I can move that.  I am not sure what the best place would be but I'll make the suggested change.  If it doesn't look good or doesn't factor well, we can always move it again.

jhogberg commented on erts/emulator/asmjit/core/jitallocator.cpp:
```diff
@@ -521,14 +521,17 @@ static Error JitAllocatorImpl_newBlock(JitAllocatorPrivateImpl* impl, JitAllocat
     bitWords = static_cast<BitWord*>(::malloc(size_t(numBitWords) * 2 * sizeof(BitWord)));
 
+  VirtMem::MemoryFlags memoryFlags = VirtMem::MemoryFlags::kAccessRWX;
+  if (Support::test(impl->options, JitAllocatorOptions::kUseLargePages))
```
This differs from `asmjit` upstream and will get clobbered as soon as we update it. In the PR you made upstream, the maintainer [asked you](https://github.com/asmjit/asmjit/pull/378#issuecomment-1712799408) to try their version of large pages but you didn't respond (in public at least).

Are there any problems with their implementation?

lexprfuncall replied:
Yes, the implementation does not work.

This is an unfortunate situation.  I contributed the change a year ago and the `asmjit` maintainer and I worked on adjusting the pull request to his liking with various functional and cosmetic changes.  Alas, in the end he would not accept it on the master branch.  Three months ago he picked out a few parts of my change and rewrote them for whatever reason without preserving the original behavior.

I am not sure why things have gone in this direction.   I can follow-up on the pull request and see if he can share his intentions for my pull request.

jhogberg replied:
> I am not sure why things have gone in this direction. I can follow-up on the pull request and see if he can share his intentions for my pull request.

I think that'd be a great help, I get the feeling you've been talking past each other.

mikpe commented:
I've only skimmed this PR, but FWIW I tried using Linux hugepages back around R15/R16 or so. Transparent hugepages were unusable (caused kernel lockups in the RHEL kernels we had at the time). What I ended up doing was to use explicit hugepages for the super carrier, but nothing else. That worked but wasn't a win (and required non-default kernel options) so I didn't pursue it.

lexprfuncall commented on erts/emulator/beam/erl_init.c:
```diff
@@ -762,6 +806,110 @@ static void ethr_ll_free(void *ptr)
+static int
+maybe_madvise_hugepage(char *from, char *to)
+{
+    extern char etext; /* &etext is the first address above the .text segment */
```
https://man7.org/linux/man-pages/man3/end.3.html

jhogberg replied:
Thanks, it would be nice to have that in the comment though. :)

okeuday replied:
While `etext`, `edata` and `end` have traditionally existed, their presence is not standardized, so it can lead to a portability problem.

jhogberg replied:
Yeah, we can always try hiding it all behind a configure test though.

lexprfuncall commented on erts/emulator/beam/erl_init.c:
```diff
@@ -762,6 +806,110 @@ static void ethr_ll_free(void *ptr)
+     * allocate 2MiB pages immediately, memory regions must be 2MiB
+     * aligned.
+     *
+     * The build should have instructed the linker to 2MiB align
+     * segments, if they are not, something has gone wrong.
```
The `etext` symbol is provided by the linker but you are right about there being no guarantee of 2MiB section alignment.  That is my oversight, I pass a flag to the linker to increase the alignment granularity but I did not include that in this change.

Here is what things probably look like by default, using a simple "hello, world" program on a vanilla Linux distribution

```
fedora {25} cc -o hello-4kib hello.c
fedora {26} readelf -l hello-4kib 

Elf file type is EXEC (Executable file)
Entry point 0x401040
There are 13 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x00000000000002d8 0x00000000000002d8  R      0x8
  INTERP         0x0000000000000318 0x0000000000400318 0x0000000000400318
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000000508 0x0000000000000508  R      0x1000
  LOAD           0x0000000000001000 0x0000000000401000 0x0000000000401000
                 0x0000000000000161 0x0000000000000161  R E    0x1000
...
```

And here is what they look like when we override the page size

```
fedora {27} cc -o hello-2mib -Wl,-zcommon-page-size=2097152 hello.c
fedora {28} readelf -l hello-2mib 

Elf file type is EXEC (Executable file)
Entry point 0x600040
There are 13 program headers, starting at offset 64

Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  PHDR           0x0000000000000040 0x0000000000400040 0x0000000000400040
                 0x00000000000002d8 0x00000000000002d8  R      0x8
  INTERP         0x0000000000000318 0x0000000000400318 0x0000000000400318
                 0x000000000000001c 0x000000000000001c  R      0x1
      [Requesting program interpreter: /lib64/ld-linux-x86-64.so.2]
  LOAD           0x0000000000000000 0x0000000000400000 0x0000000000400000
                 0x0000000000000508 0x0000000000000508  R      0x1000
  LOAD           0x0000000000200000 0x0000000000600000 0x0000000000600000
                 0x0000000000000161 0x0000000000000161  R E    0x1000
...
```

You can see that the text segment starts at 0x60000 instead of 0x401000, which is what we need.  This flag is something that can probably be added by the `configure` script contingent on all of the other THP stuff being present.

lexprfuncall commented on erts/emulator/beam/erl_init.c:
```diff
@@ -793,6 +941,9 @@ early_init(int *argc, char **argv) /*
     char envbuf[21]; /* enough for any 64-bit integer */
     size_t envbufsz;
     int skip_read_topology = 0;
+#if defined(__linux__) && defined(__x86_64__) && defined(__LP64__)
```
Sure, I can move this into `configure`.  This is limited to 64-bit x86 since 32-bit x86 has a 4MiB alternate page size, inherited from the i860, which you can read about [here](http://www.rcollins.org/articles/4mpages/4MOverview.html).

Outside of x86, the sizes of the alternate page sizes varies a lot.  64-bit ARM on Linux has 4KiB 2MiB and 1GiB pages like the x86.  32-bit ARM has, IIRC, 1MiB pages.  64-bit ARM with a 64KiB page has 512MiB pages.  PowerPC has 64KiB, 16MiB, pages, at least.  It's a zoo.

lexprfuncall commented on erts/emulator/beam/erl_init.c:
```diff
@@ -900,6 +1051,16 @@ early_init(int *argc, char **argv) /*
 		    break;
 		}
+		case 'h':
+		    if (has_prefix("hugify_self", argv[i] + 1)) {
```
No downside, I think.  I suspect that kernels which support the feature that makes this change possible are widespread now.  But even if you don't have the feature, the system call is advisory and nothing catastrophic will happen.

Now, before the patches to Linux to use THP for file mappings were upstreamed, there were all sorts of hacks in use for having `.text` on 2MiB pages.  They'd either do the work of the loader or save aside and unmap the `.text` segment live and then remap it with 2MiB pages.  I think this won't break that stuff.  But, I suspect that if you were smart enough to make those things work, you are probably smart enough to work around any possible issues with this change.

lexprfuncall commented on erts/emulator/beam/erl_init.c:
```diff
@@ -1763,6 +1932,30 @@ erl_start(int argc, char **argv)
 #endif
         break;
 
+        case 'j':
+#ifdef BEAMASM
+	    if (maybe_get_bool_arg("jit_use_thp", argv, &i, &erts_jit_use_thp)) {
+#ifndef __linux__
+                erts_fprintf(stderr, "%s is not supported on this platform\n", argv[i - 1]);
+                erts_usage();
+#endif
+                break;
+            }
+            if (maybe_get_bool_arg("jit_use_hugetlb", argv, &i, &erts_jit_use_hugetlb)) {
+#ifndef __linux__
+                erts_fprintf(stderr, "%s is not supported on this platform\n", argv[i - 1]);
+                erts_usage();
+#endif
+                break;
+            }
+#else
+            erts_fprintf(stderr,
+                         "JIT is not supported on this system (option %s)\n",
+                         argv[i]);
+            erts_usage();
+#endif
+        break;
```
Sounds good.  I am happy to fix that.  Presumably, we'd also want the `#ifdef __linux__` to be replaced with something more descriptive defined by the `configure` script?

jhogberg replied:
Yes please :)

lexprfuncall commented on erts/emulator/sys/common/erl_mmap.c:
```diff
@@ -2246,13 +2296,25 @@ erts_mmap_init(ErtsMemMapper* mm, ErtsMMapInit *init)
 	     * reserved all the time.
 	     */
-	    start = os_mmap(NULL, sz, 1);
+	    UWord alignment = init->use_thp ? 2*1024*1024 : ERTS_SUPERALIGNED_SIZE;
```
Thanks, that sounds better.

RoadRunnr replied:
Wouldn't it make sense, at least on  linux, to read that value from `/sys/kernel/mm/transparent_hugepage/hpage_pmd_size` ?

lexprfuncall commented:
@jhogberg you left some other comments which all seem very reasonable.  I'll follow-up on them after the weekend.  Also, sorry for all the one-off comments.  Responding from an e-mail does not give you the "start a review" option which lead me to having my replies dribble out incrementally.  I'll be more tidy with the remainder of my comments.

jhogberg commented:
> @jhogberg you left some other comments which all seem very reasonable. I'll follow-up on them after the weekend.

Thanks!

> Also, sorry for all the one-off comments. Responding from an e-mail does not give you the "start a review" option which lead me to having my replies dribble out incrementally. I'll be more tidy with the remainder of my comments.

I don't mind. :)

lexprfuncall commented:
@mikpe we have had pretty good luck even back in the R15/R16 era with large pages.  Back then, it was shown to be profitable to align allocations on a 2MiB and use FreeBSD's superpage feature on 64-bit x86.

As for HugeTLB, I have an option to use them for `asmjit`'s dual-mapping mode on Linux where THP is not an option.  In general, my experience is that the administrative overhead for having large amounts of huge pages is tricky when you are running in a shared environment and rebooting the machine between jobs is not an option.  For the JIT where we use, say, 128MiB of memory, it's not a big deal to reserve a few 2MiB pages everywhere.  With today's memory sizes, that will fly under the radar.  For the super carrier which can be 10s or 100s of GiB, it's less practical.

lexprfuncall reviewed (commented):
Thank you everyone for all of your feedback.  I think I've made all of the requested changes.  One significant highlight is that I have pulled out the JIT-related changes.  Because of its entanglement with third-party code, I will create a separate pull request for it.

jhogberg commented:
Thanks, I've added it to our daily builds. :)

lexprfuncall reviewed (commented):
Took a few tries but the CI is now green.  The combination of an older version of `clang` and the `-Werror` and `-Wunused-command-line-argument` options ultimately meant some extra code was needed to find the subset of the flags clang-10 supports for segment alignment among those GCC actually uses.

jhogberg commented on erts/emulator/sys/unix/sys.c:
```diff
@@ -230,6 +273,49 @@ thr_create_cleanup(void *vtcdp)
+            continue;
+        if (from > &etext)
+            break;
+        if ((UWord)from % sys_large_page_size != 0)
```
This crashes if `is_linux_thp_enabled()` returns non-zero and `/sys/kernel/mm/transparent_hugepage/hpage_pmd_size` cannot be opened, resulting in `sys_large_page_size` being zero. We've got two test rigs where this is the case, one with kernel version `4.4.74` and the other with `3.12.60`.

Perhaps `is_linux_thp_enabled` should also check `get_large_page_size() != 0`?

lexprfuncall replied:
Sorry, I think I let that bug creep in after I rearranged the code.

Along the lines of what you have suggested, I have added an early exit in this function if `sys_large_page_size == 0`.  That should prevent the later `% sys_large_page_size` from exploding.  This variation appealed to my aesthetic sense of having the thp enabled check just be about reading and parsing the relevant `sysfs(5)` file.  However, if you think it's better to put this check in `is_linux_thp_enabled` I am certainly okay with that too

jhogberg commented:
This is pretty much ready to merge now. I think the only thing left is to move the flag affecting `mseg_alloc` over to https://www.erlang.org/doc/man/erts_alloc.html#system-flags-effecting-erts_alloc and documenting it.

Since we may want to introduce explicit huge pages later (optionally with size when arch supports that?), I'm thinking of calling the option `+MMhp transparent | off` but anyone has a better suggestion I'm all ears.

lexprfuncall commented:
> This is pretty much ready to merge now. I think the only thing left is to move the flag affecting `mseg_alloc` over to https://www.erlang.org/doc/man/erts_alloc.html#system-flags-effecting-erts_alloc and documenting it.

Sure.  Would you like me to take care of that?

> Since we may want to introduce explicit huge pages later (optionally with size when arch supports that?), I'm thinking of calling the option `+MMhp transparent | off` but anyone has a better suggestion I'm all ears.

I think my choice of a flag name encoding a Linux-specific mechanism like THP or HugeTLB was unfortunate as the patch is trivially generalized to other operating systems which have semantics slightly different from Linux THP.  If I had to do it again, I'd call it `+mseg_use_large_pages` or an unpronounceable variant thereof.

Also, prescribing the mechanism on Linux in some all-or-nothing way prevents using both mechanisms where appropriate.  For example, 1GiB pages are not supported by the iTLB so we like to use HugeTLB to get a few 1GiB pages for data and THP for the rest.

Anyway, what I have in mind is having the "on" value enable some sensible default support for large pages.  THP is a good default for Linux and we can support FreeBSD, macOS, Solaris and Win32 in a similar way with a very small amount of new platform specific code (likely less than was required for Linux).  More exotic things like Linux HugeTLB or the similar feature on AIX probably require a few extra flags to specify policies for things like selecting the page size, hoarding the reserved pages at startup, or what to do when the reserved pages are exhausted.  That will be hard to configure in a single flag so I’d imagine it would be kindest to users to configure that separately.

What do you think?  As an aside, I’m happy to make the follow-up pull request to support FreeBSD and macOS, at least.

vans163 commented:
> > This is pretty much ready to merge now. I think the only thing left is to move the flag affecting `mseg_alloc` over to https://www.erlang.org/doc/man/erts_alloc.html#system-flags-effecting-erts_alloc and documenting it.
> 
> Sure. Would you like me to take care of that?
> 
> > Since we may want to introduce explicit huge pages later (optionally with size when arch supports that?), I'm thinking of calling the option `+MMhp transparent | off` but anyone has a better suggestion I'm all ears.
> 
> I think my choice of a flag name encoding a Linux-specific mechanism like THP or HugeTLB was unfortunate as the patch is trivially generalized to other operating systems which have semantics slightly different from Linux THP. If I had to do it again, I'd call it `+mseg_use_large_pages` or an unpronounceable variant thereof.
> 
> Also, prescribing the mechanism on Linux in some all-or-nothing way prevents using both mechanisms where appropriate. For example, 1GiB pages are not supported by the iTLB so we like to use HugeTLB to get a few 1GiB pages for data and THP for the rest.
> 
> Anyway, what I have in mind is having the "on" value enable some sensible default support for large pages. THP is a good default for Linux and we can support FreeBSD, macOS, Solaris and Win32 in a similar way with a very small amount of new platform specific code (likely less than was required for Linux). More exotic things like Linux HugeTLB or the similar feature on AIX probably require a few extra flags to specify policies for things like selecting the page size, hoarding the reserved pages at startup, or what to do when the reserved pages are exhausted. That will be hard to configure in a single flag so I’d imagine it would be kindest to users to configure that separately.
> 
> What do you think? As an aside, I’m happy to make the follow-up pull request to support FreeBSD and macOS, at least.

This would be really nice, we have long wanted hugepages to work natively on Linux (seems they are only supported on FreeBSD via superpages).

lexprfuncall commented:
> Yes please. :)

Sure, I'll take care of it in the next few days.

> We can always have +MMhp<> more specific, for example +MMhpi for instructions (.text segment, JIT once supported) and +MMhpd for data.

What I was proposing might be a little different, which is to have the flag toggle the use of large pages and leave it up to the runtime to pick a mechanism, rather than specifying the details on the commandline of what mechanism (transparent, explicit, etc.) and where (static text, data, etc.)

Consider the following cases...

1) For FreeBSD, HP-UX, Linux with THP set to `always`, recent Solaris and, supposedly, recent Windows, you can take advantage of large pages by, at most, arranging to allocate virtual memory at the right alignment and size.

2) For Linux with THP set to `madvise`, MacOS, older Solaris, and most or all Windows, there are advisory and mandatory mechanisms available for requesting large pages.  If the mechanism is mandatory, you need to retry it and fall back to something reasonable.

3) For AIX and Linux with HugeTLB, the mechanism allocates from a finite pool of reserved pages.  This also creates lots of corner cases (consider the case of calling `fork(2)`) so it can create correctness issues if not used with care.

At least for cases 1 and 2, it is enough to have a single flag that turns large pages on or off through whatever combination of mechanisms makes sense on the host operating system.  This is easiest for users and this flag would be a good candidate for eventually being on by default since there is rarely a downside to having it enabled.

For case 3, a hairy parameter, like the `+MMhp<>` you proposed, that can encode all of the information needed to configure the huge pages (page sizes, count of pages, NUMA domain, etc.) and what to do if the pages are not available is probably unavoidable.  For users that have already gone to the trouble to provision huge pages, the added burden is probably acceptable.

> The 1GiB pages for data case could be handled through the super-carrier mechanism (introduce +MMschp?), which already handles much of what you mentioned with regards to reservation, running out of pages, and so on.

I also implemented support for 1GiB pages using HugeTLB.  It was a more invasive change since you need to decide which parts of the super carrier get the 1GiB huge pages.  (I also had to replace a call to `fork(2)`.)  If the super carrier runs out of 1GiB pages it is still reasonable to use other mechanisms.  IIRC, the fallback in the super carrier is either mseg or sysalloc which is less efficient, at least for us.

lexprfuncall commented:
> Then we're on the same page, you shouldn't need to touch the tricky flags unless you need to, but they should be there for "case 3" once we cross that bridge.

To confirm, the direction is to rename the flag as you had suggested above but have an "on" and "off" setting for now, with off as the default, right?

> Then we'll fall back differently under +MMschp on or whatever we decide to call it, it doesn't have to work the way it does today :-)

Indeed, this will need a different design.

lexprfuncall commented:
> Yeah, `on`/`off` works :-)

Thanks for clarifying.  I will go ahead update the flag and add the documentation sometime next week.

lexprfuncall commented:
@jhogberg I have updated the documentation and changed the flag name to be a `+M` option that can be set to either `on` or `off` with `off` as the default.  Since it turns out "huge page" means something specific on Windows I just used the large page terminology throughout.  Please let me know if there are any other needed changes.
